File: tree/binary_search_tree/98_validate_binary_search_tree.cc
#include <vector>
#include <climits>

#include "../tree_node.h"

// 2024/01/09: 自己实现了中序遍历版本.

bool IsValidBST(TreeNode* root) {
  std::vector<int> inordered_traversal;
  inorderedTraversal(inordered_traversal, root);
  for (int i = 0; i < inordered_traversal.size() - 1; ++i) {
    if (inordered_traversal[i] >= inordered_traversal[i + 1]) {
      return false;
    }
  }
  return true;
}

void inorderedTraversal(std::vector<int>& inordered_traversal, TreeNode* root) {
  if (root == nullptr) {
    return;
  }
  inorderedTraversal(inordered_traversal, root->left);
  inordered_traversal.push_back(root->val);
  inorderedTraversal(inordered_traversal, root->right);
}

// 2024/01/09: 递归版本, 直接看的答案
bool Helper(TreeNode* root, long long lower, long long upper) {
  if (root == nullptr) {
    return true;
  }
  if (root->val <= lower || root->val >= upper) {
    return false;
  }
  return Helper(root->left, lower, root->val) &&
         Helper(root->right, root->val, upper);
}

bool IsValidBST(TreeNode* root) { return Helper(root, LONG_MIN, LONG_MAX); }

File: tree/binary_search_tree/98_validate_binary_search_tree.py
from tree.tree_node import TreeNode, create_tree_node
from typing import Optional, Tuple

# 2026/05/18: 自己实现直接递归版本, 也可以中序遍历后看得到的数组 List 是否严格升序.
flag = True


def get_min_max(root: Optional[TreeNode]) -> Tuple[int, int]:
    global flag
    if root is None:
        return float("inf"), float("-inf")

    value = root.val
    left_min_value, left_max_value = get_min_max(root.left)
    right_min_value, right_max_value = get_min_max(root.right)
    if value <= left_max_value or value >= right_min_value:
        flag = False
    return min(value, left_min_value, right_min_value), max(
        value, left_max_value, right_max_value
    )


def is_valid_bst(root: Optional[TreeNode]) -> bool:
    get_min_max(root)
    return flag


root = [5, 1, 4, None, None, 3, 6]
root = create_tree_node(root)
print(is_valid_bst(root))
